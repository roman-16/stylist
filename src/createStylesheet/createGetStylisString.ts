import { Breakpoints, Styles, Theme, ThemeProperty, UtilsObject } from '@/types';

const hyphenateRegex = /[A-Z]|^ms/g;
const resolveKey = (styleName: string) => styleName.replace(hyphenateRegex, '-$&').toLowerCase();

interface Scales {
  keys: string[];
  scale: ThemeProperty;
}

const resolveScales = (key: string, value: string | number, scales: Scales[]): string => {
  const getStringValue = () => {
    if (typeof value === 'number') throw new Error(`Key '${key}' can't have value '${value}' of type number.`);

    return value;
  };

  for (let i = 0; i < scales.length; i++) {
    const { keys, scale } = scales[i];

    if (keys.includes(key)) {
      if (
        (Array.isArray(scale) && typeof value === 'number') ||
        (!Array.isArray(scale) && typeof scale === 'object' && scale !== null)
      ) {
        // FIXME: Here are weird errors with the value on array or object, I don't know how to fix this yet so this must do for now
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unsafe-return
        return scale[value] ?? getStringValue();
      }

      return getStringValue();
    }
  }

  return getStringValue();
};

const ownUtils: UtilsObject = {
  m: (value) => ({ margin: value }),
  mt: (value) => ({ marginTop: value }),
  mr: (value) => ({ marginRight: value }),
  mb: (value) => ({ marginBottom: value }),
  ml: (value) => ({ marginLeft: value }),
  mx: (value) => ({ marginRight: value, marginLeft: value }),
  my: (value) => ({ marginTop: value, marginBottom: value }),

  p: (value) => ({ padding: value }),
  pt: (value) => ({ paddingTop: value }),
  pr: (value) => ({ paddingRight: value }),
  pb: (value) => ({ paddingBottom: value }),
  pl: (value) => ({ paddingLeft: value }),
  px: (value) => ({ paddingRight: value, paddingLeft: value }),
  py: (value) => ({ paddingTop: value, paddingBottom: value }),

  bg: (value) => ({ backgroundColor: value }),
};

const createGetStylisString = ({
  breakpoints,
  theme,
  utilsObject,
}: {
  breakpoints?: Breakpoints;
  theme?: Theme;
  utilsObject?: UtilsObject;
}) => {
  const mergedUtilsObject: UtilsObject = { ...utilsObject, ...ownUtils };
  const scales: Scales[] = [];

  if (theme?.space) {
    scales.push({
      keys: [
        'margin',
        'marginTop',
        'marginRight',
        'marginBottom',
        'marginLeft',
        'padding',
        'paddingTop',
        'paddingRight',
        'paddingBottom',
        'paddingLeft',
        'gap',
        'gridGap',
        'columnGap',
        'gridColumnGap',
        'MozColumnGap',
        'WebkitColumnGap',
        'rowGap',
        'gridRowGap',

        // Own utils
        'm',
        'mt',
        'mr',
        'mb',
        'ml',
        'mx',
        'my',

        'p',
        'pt',
        'pr',
        'pb',
        'pl',
        'px',
        'py',
      ],
      scale: theme.space,
    });
  }

  if (theme?.fontSizes) {
    scales.push({ keys: ['fontSize'], scale: theme.fontSizes });
  }

  if (theme?.colors) {
    // 'bg' is from own utils
    scales.push({ keys: ['color', 'backgroundColor', 'borderColor', 'bg'], scale: theme.colors });
  }

  if (theme?.fonts) {
    scales.push({ keys: ['fontFamily'], scale: theme.fonts });
  }

  if (theme?.fontWeights) {
    scales.push({ keys: ['fontWeight'], scale: theme.fontWeights });
  }

  if (theme?.lineHeights) {
    scales.push({ keys: ['lineHeight'], scale: theme.lineHeights });
  }

  if (theme?.letterSpacings) {
    scales.push({ keys: ['letterSpacing'], scale: theme.letterSpacings });
  }

  if (theme?.sizes) {
    scales.push({ keys: ['width', 'height', 'minWidth', 'maxWidth', 'minHeight', 'maxHeight'], scale: theme.sizes });
  }

  if (theme?.borders) {
    scales.push({ keys: ['border', 'borderTop', 'borderRight', 'borderBottom', 'borderLeft'], scale: theme.borders });
  }

  if (theme?.borderWidths) {
    scales.push({ keys: ['borderWidth'], scale: theme.borderWidths });
  }

  if (theme?.borderStyles) {
    scales.push({ keys: ['borderStyle'], scale: theme.borderStyles });
  }

  if (theme?.radii) {
    scales.push({ keys: ['borderRadius'], scale: theme.radii });
  }

  if (theme?.shadows) {
    scales.push({ keys: ['boxShadow', 'textShadow'], scale: theme.shadows });
  }

  if (theme?.zIndices) {
    scales.push({ keys: ['zIndex'], scale: theme.zIndices });
  }

  if (theme?.transitions) {
    scales.push({ keys: ['transition'], scale: theme.transitions });
  }

  const resolveValue = (key: string, value: string) => resolveScales(key, value, scales);

  const getStylisString = (styles: Styles, disableUtil = false): string =>
    Object.entries(styles).reduce((previous, [key, value]) => {
      const breakpoint = breakpoints?.[key];
      const util = mergedUtilsObject?.[key];
      const resolvedKey = resolveKey(key);

      if (breakpoint) {
        const stylisString = getStylisString(value);

        return `${previous}${breakpoint(stylisString)}`;
      }

      if (!disableUtil && util) {
        const utilValue = util(value);

        if (typeof utilValue === 'object') {
          return `${previous}${getStylisString(utilValue, true)}`;
        }

        return `${previous}${resolvedKey}:${resolveValue(key, utilValue ?? value)};`;
      }

      if (typeof value === 'object') {
        const processedValue = getStylisString(value);

        return `${previous}${resolvedKey}{${processedValue}}`;
      }

      return `${previous}${resolvedKey}:${resolveValue(key, value)};`;
    }, '');

  return getStylisString;
};

export default createGetStylisString;
