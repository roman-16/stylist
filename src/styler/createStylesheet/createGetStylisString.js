const hyphenateRegex = /[A-Z]|^ms/g;
const resolveKey = (styleName) => styleName.replace(hyphenateRegex, '-$&').toLowerCase();

const resolveScales = (key, value, scales) => {
  for (let i = 0; i < scales.length; i++) {
    const { keys, scale } = scales[i];

    if (keys.includes(key)) {
      if (
        (Array.isArray(scale) && typeof value === 'number') ||
        (typeof scale === 'object' && scale !== null && !Array.isArray(scale))
      ) {
        return scale[value] ?? value;
      }

      return value;
    }
  }

  return value;
};

export default ({ breakpoints = {}, theme = {}, utils = {} } = {}) => {
  const resolveValue = (key, value) =>
    resolveScales(key, value, [
      {
        keys: [
          'margin',
          'marginTop',
          'marginRight',
          'marginBottom',
          'marginLeft',
          'padding',
          'paddingTop',
          'paddingRight',
          'paddingBottom',
          'paddingLeft',
          'gridGap',
          'gridColumnGap',
          'gridRowGap',
        ],
        scale: theme.space,
      },
      { keys: ['fontSize'], scale: theme.fontSizes },
      { keys: ['color', 'backgroundColor', 'borderColor'], scale: theme.colors },
      { keys: ['fontFamily'], scale: theme.fonts },
      { keys: ['fontWeight'], scale: theme.fontWeights },
      { keys: ['lineHeight'], scale: theme.lineHeights },
      { keys: ['letterSpacing'], scale: theme.letterSpacings },
      { keys: ['width', 'height', 'minWidth', 'maxWidth', 'minHeight', 'maxHeight'], scale: theme.sizes },
      { keys: ['border', 'borderTop', 'borderRight', 'borderBottom', 'borderLeft'], scale: theme.borders },
      { keys: ['borderWidth'], scale: theme.borderWidths },
      { keys: ['borderStyle'], scale: theme.borderStyles },
      { keys: ['borderRadius'], scale: theme.radii },
      { keys: ['boxShadow', 'textShadow'], scale: theme.shadows },
      { keys: ['zIndex'], scale: theme.zIndices },
      { keys: ['transition'], scale: theme.transitions },
    ]);

  const getStylisString = (styles, disableUtil = false) =>
    Object.entries(styles).reduce((previous, [key, value]) => {
      const breakpoint = breakpoints[key];
      const util = utils[key];
      const resolvedKey = resolveKey(key);

      if (breakpoint) {
        const stylisString = getStylisString(value);

        return `${previous}${breakpoint(stylisString)}`;
      }

      if (!disableUtil && util) {
        const utilValue = util(value);

        if (typeof utilValue === 'object') {
          return `${previous}${getStylisString(utilValue, true)}`;
        }

        return `${previous}${resolvedKey}:${resolveValue(key, utilValue ?? value)};`;
      }

      if (typeof value === 'object') {
        const processedValue = getStylisString(value);

        return `${previous}${resolvedKey}{${processedValue}}`;
      }

      return `${previous}${resolvedKey}:${resolveValue(key, value)};`;
    }, '');

  return getStylisString;
};
