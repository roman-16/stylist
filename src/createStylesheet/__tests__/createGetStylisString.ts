import createGetStylisString from '../createGetStylisString';

const getStylisString = createGetStylisString({
  theme: {
    space: ['2px', '4px'],
    fontSizes: ['16px', '32px'],
    colors: { $red: '#FF0000', $blue: '#0000FF' },
    fonts: { $h: 'Arial', $t: 'Helvetica' },
    fontWeights: ['300', '500'],
    lineHeights: ['1.6', '2.2'],
    letterSpacings: ['1px', '2px'],
    sizes: ['32px', '64px'],
    borders: { $thin: '1px solid black', $thicc: '4px solid black' },
    borderWidths: ['1px', '4px'],
    borderStyles: { $normal: 'solid', $link: 'dotted' },
    radii: ['8px', '16px'],
    shadows: { $small: '1px 1px', $big: '4px 4px' },
    zIndices: ['0', '1'],
    transitions: { $fast: 'all 1s', $slow: 'all 4s' },
  },
  utilsObject: {
    display: (value) => (value === 'inline-block' ? 'block' : value),
  },
  breakpoints: {
    $small: (rules) => `@media (min-width: 576px) { ${rules} }`,
    $medium: (rules) => `@media (min-width: 768px) { ${rules} }`,
  },
});

it('resolves the basic utils', () => {
  expect(
    getStylisString({
      m: 0,
      p: 1,
      bg: '$red',
    }),
  ).toMatchSnapshot();

  expect(
    getStylisString({
      mt: 0,
      mr: 1,
      mb: 0,
      ml: 1,
      pt: 0,
      pr: 1,
      pb: 0,
      pl: 1,
    }),
  ).toMatchSnapshot();

  expect(
    getStylisString({
      mx: 0,
      my: 1,
      px: 0,
      py: 1,
    }),
  ).toMatchSnapshot();
});

it('allows styling', () => {
  expect(
    getStylisString({
      margin: 0,
      fontSize: 1,
      backgroundColor: '$red',
      fontFamily: '$t',
      fontWeight: 0,
      lineHeight: 1,
      letterSpacing: 0,
      height: 1,
      border: '$thin',
      borderWidth: 0,
      borderStyle: '$link',
      borderRadius: 1,
      boxShadow: '$small',
      zIndex: 0,
      transition: '$slow',

      padding: '7px', // Normal style
      p: 1, // Util style with theme
      display: 'inline-block', // Custom util

      // Breakpoints
      $small: {
        margin: 1,
      },
    }),
  ).toMatchSnapshot();
});

it("throws an error if it can't resolve a scale", () => {
  expect(() =>
    getStylisString({
      margin: 2,
    }),
  ).toThrow();
});
