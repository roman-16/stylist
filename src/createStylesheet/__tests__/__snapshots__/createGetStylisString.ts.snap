// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows styling 1`] = `"margin:2px;font-size:32px;background-color:#FF0000;font-family:Helvetica;font-weight:300;line-height:2.2;letter-spacing:1px;height:64px;border:1px solid black;border-width:1px;border-style:dotted;border-radius:16px;box-shadow:1px 1px;z-index:0;transition:all 4s;padding:7px;padding:4px;display:block;@media (min-width: 576px) { margin:4px; }"`;

exports[`resolves the basic utils 1`] = `"margin:2px;padding:4px;background-color:#FF0000;"`;

exports[`resolves the basic utils 2`] = `"margin-top:2px;margin-right:4px;margin-bottom:2px;margin-left:4px;padding-top:2px;padding-right:4px;padding-bottom:2px;padding-left:4px;"`;

exports[`resolves the basic utils 3`] = `"margin-right:2px;margin-left:2px;margin-top:4px;margin-bottom:4px;padding-right:2px;padding-left:2px;padding-top:4px;padding-bottom:4px;"`;
